.github/workflows/breaking-change-warning.yml

```diff
@@ -0,0 +1,45 @@
+name: Breaking Change Warning
+on:
+  pull_request:
+    types: [ opened, synchronize, reopened ]
+    branches:
+      - main
+      - release/*
+      - next
+
+jobs:
+  check_broken_types:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout repo
+        uses: actions/checkout@2541b1294d2704b0964813337f33b291d3f8596b # pin@v3
+        with:
+          fetch-depth: 0
+
+      - name: Check for changes to broken types
+        id: check
+        run: |
+          declare -a arr=($(git diff --name-only --diff-filter=ACMRT ${{ github.event.pull_request.base.sha }} ${{ github.sha }} | xargs))
+          for i in "${arr[@]}"; do
+            if [[ ! "$i" =~ package.json$ ]]; then
+              continue
+            fi
+
+            git checkout ${{ github.head_ref }} --quiet
+            pr=$(jq .typeValidation.broken "$i" | tr -d '\n')
+
+            git checkout ${{ github.base_ref }} --quiet
+            base=$(jq .typeValidation.broken "$i" | tr -d '\n')
+
+            if [ "$pr" != "$base" ]; then
+              echo '::set-output name=isBrokenDetected::true'
+              break
+            fi
+          done
```

Comment:
I highly recommend moving this logic outside of the workflow. Think of the workflow as an "orchestrator" only. It shouldn't have any "real" logic in it -- rather, it should just wire things together, similar to how you might wire things together manually.

So with that in mind, you could break this down into two pieces.

1. Given a list of files, the program checks those files for relevant changes.
2. Determine the list of files to check.

The first part should be external to the workflow -- think of it as a function the workflow calls. One approach would be to make this a CLI command that is installed and run as part of the workflow. Another is to make it a JS file that you run directly using node. I think the CLI approach is more robust but does mean a bit more infrastructure.

For the second part, it could live in the workflow if needed. You could list the changed files using git and pass that in via stdin to the CLI command.

The big advantage of moving this out of the workflow is testing. You can test the core logic of the workflow without running the action. @sonalivdeshpande may have additional thoughts because she has hit some of these problems with the branch merge workflow.